File: src/main/java/me/testaccount666/serversystem/managers/messages/MessageManager.java
package me.testaccount666.serversystem.managers.messages;

import lombok.Getter;
import lombok.SneakyThrows;
import lombok.experimental.UtilityClass;
import me.testaccount666.serversystem.ServerSystem;
import me.testaccount666.serversystem.managers.PlaceholderManager;
import me.testaccount666.serversystem.managers.config.ConfigReader;
import me.testaccount666.serversystem.managers.config.ConfigurationManager;
import me.testaccount666.serversystem.userdata.User;
import me.testaccount666.serversystem.utils.ChatColor;
import me.testaccount666.serversystem.utils.ComponentColor;
import net.kyori.adventure.text.Component;

import javax.annotation.Nullable;
import java.io.FileNotFoundException;
import java.util.Optional;
import java.util.logging.Level;

@UtilityClass
public class MessageManager {
    public static final String FALLBACK_LANGUAGE = "english";
    @Getter
    private static LanguageLoader _LanguageLoader;
    private static PlaceholderManager _PlaceholderManager;
    @Getter
    private static String _DefaultLanguage;

    public static void initialize() throws FileNotFoundException {
        _PlaceholderManager = new PlaceholderManager();
        _LanguageLoader = new LanguageLoader();
        _DefaultLanguage = ServerSystem.Instance.getRegistry().getService(ConfigurationManager.class).getGeneralConfig().getString("Language.DefaultLanguage", FALLBACK_LANGUAGE);

        var messageOptional = getLanguageLoader().getMessageReader(_DefaultLanguage);
        var mappingsOptional = getLanguageLoader().getMappingReader(_DefaultLanguage);

        if (messageOptional.isPresent() && mappingsOptional.isPresent()) return;
        ServerSystem.getLog().warning("Could not load default language '${_DefaultLanguage}'! Falling back to '${FALLBACK_LANGUAGE}'...");
        ServerSystem.getLog().warning("Make sure you have a 'messages.yml' AND 'mappings.yml' file in your language folder!");
        _DefaultLanguage = FALLBACK_LANGUAGE;
    }

    public static String formatMessage(String message, User commandSender, @Nullable String targetName, @Nullable String label, boolean addPrefix) {
        if (message == null) return "";

        if (addPrefix) {
            var prefix = getMessage(commandSender, "General.Prefix").orElse("");
            message = prefix + message;
        }

        return ChatColor.translateColor(applyPlaceholders(message, commandSender, targetName, label));
    }

    public static String applyPlaceholders(String message, User commandSender, @Nullable String targetName, @Nullable String label) {
        return _PlaceholderManager.applyPlaceholders(message, commandSender, targetName, label != null? label : "");
    }

    /**
     * Formats a message with placeholders and color codes, returning a Component.
     * This method is similar to formatMessage but returns a Component instead of a String.
     *
     * @param message       The message to format
     * @param commandSender The user who sent the command
     * @param targetName    The name of the target user (optional)
     * @param label         The command label (optional)
     * @param addPrefix     Whether to add the prefix to the message
     * @return The formatted message as a Component
     */
    public static Component formatMessageAsComponent(String message, User commandSender, @Nullable String targetName, @Nullable String label, boolean addPrefix) {
        if (message == null) return Component.empty();

        if (addPrefix) {
            var prefix = getMessage(commandSender, "General.Prefix").orElse("");
            message = prefix + message;
        }

        var processedMessage = _PlaceholderManager.applyPlaceholders(message, commandSender, targetName, label != null? label : "");
        return ComponentColor.translateToComponent(processedMessage);
    }

    /**
     * Gets a message from the messages file for the specified user's language.
     *
     * @param user        The user to get the message for
     * @param messagePath The path to the message in the messages file
     * @return An Optional containing the message, or empty if not found
     */
    @SneakyThrows
    public static Optional<String> getMessage(User user, String messagePath) {
        if (_PlaceholderManager == null) throw new IllegalStateException("MessageManager was not yet initialized. Call initialize first.");
        messagePath = "Messages.${messagePath}";

        var language = user != null? user.getPlayerLanguage() : _DefaultLanguage;


        ConfigReader reader;
        try {
            reader = getConfigReader(language);
        } catch (FileNotFoundException exception) {
            ServerSystem.getLog().log(Level.WARNING, "Failed to load messages for language '${language}': ${exception.getMessage()}", exception);
            reader = getConfigReader(_DefaultLanguage);
        }

        var message = reader.getString(messagePath, null);

        if (message == null) {
            ServerSystem.getLog().warning("Message '${messagePath}' not found for language ${language}!");
            return Optional.empty();
        }

        return Optional.of(message);
    }

    /**
     * Gets the ConfigReader for the specified language.
     *
     * @param language The language to get the ConfigReader for
     * @return The ConfigReader for the specified language
     * @throws FileNotFoundException If the messages file couldn't be found
     */
    private static ConfigReader getConfigReader(String language) throws FileNotFoundException {
        var readerOptional = _LanguageLoader.getMessageReader(language);
        return readerOptional.orElseGet(() -> _LanguageLoader.getMessageReader(_DefaultLanguage).get());
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/managers/messages/MessageManager.kt
package me.testaccount666.serversystem.managers.messages

import me.testaccount666.serversystem.ServerSystem
import me.testaccount666.serversystem.managers.PlaceholderManager
import me.testaccount666.serversystem.managers.config.ConfigReader
import me.testaccount666.serversystem.managers.config.ConfigurationManager
import me.testaccount666.serversystem.userdata.User
import me.testaccount666.serversystem.utils.ChatColor
import me.testaccount666.serversystem.utils.ComponentColor
import net.kyori.adventure.text.Component
import java.io.FileNotFoundException
import java.util.*
import java.util.function.Supplier
import java.util.logging.Level

object MessageManager {
    const val FALLBACK_LANGUAGE: String = "english"

    @JvmStatic
    lateinit var languageLoader: LanguageLoader

    @JvmStatic
    lateinit var placeholderManager: PlaceholderManager

    @JvmStatic
    var defaultLanguage: String? = null

    @JvmStatic
    fun initialize() {
        placeholderManager = PlaceholderManager()
        languageLoader = LanguageLoader()
        defaultLanguage = ServerSystem.instance.registry.getService<ConfigurationManager>().generalConfig!!
            .getString("Language.DefaultLanguage", FALLBACK_LANGUAGE)

        val messageOptional = languageLoader.getMessageReader(defaultLanguage!!)
        val mappingsOptional = languageLoader.getMappingReader(defaultLanguage!!)

        if (messageOptional.isPresent && mappingsOptional.isPresent) return
        ServerSystem.log.warning("Could not load default language '${defaultLanguage}'! Falling back to '${FALLBACK_LANGUAGE}'...")
        ServerSystem.log.warning("Make sure you have a 'messages.yml' AND 'mappings.yml' file in your language folder!")
        defaultLanguage = FALLBACK_LANGUAGE
    }

    @JvmStatic
    fun formatMessage(message: String?, commandSender: User, targetName: String?, label: String?, addPrefix: Boolean): String {
        var message = message ?: return ""

        if (addPrefix) {
            val prefix = getMessage(commandSender, "General.Prefix").orElse("")
            message = prefix + message
        }

        return ChatColor.translateColor(applyPlaceholders(message, commandSender, targetName, label))
    }

    @JvmStatic
    fun applyPlaceholders(message: String, commandSender: User, targetName: String?, label: String?): String {
        return placeholderManager.applyPlaceholders(message, commandSender, targetName, label ?: "")
    }

    /**
     * Formats a message with placeholders and color codes, returning a Component.
     * This method is similar to formatMessage but returns a Component instead of a String.
     *
     * @param message       The message to format
     * @param commandSender The user who sent the command
     * @param targetName    The name of the target user (optional)
     * @param label         The command label (optional)
     * @param addPrefix     Whether to add the prefix to the message
     * @return The formatted message as a Component
     */
    @JvmStatic
    fun formatMessageAsComponent(message: String?, commandSender: User, targetName: String?, label: String?, addPrefix: Boolean): Component {
        var message = message ?: return Component.empty()

        if (addPrefix) {
            val prefix = getMessage(commandSender, "General.Prefix").orElse("")
            message = prefix + message
        }

        val processedMessage = placeholderManager.applyPlaceholders(message, commandSender, targetName, label ?: "")
        return ComponentColor.translateToComponent(processedMessage)
    }

    /**
     * Gets a message from the messages file for the specified user's language.
     *
     * @param user        The user to get the message for
     * @param messagePath The path to the message in the messages file
     * @return An Optional containing the message, or empty if not found
     */
    @JvmStatic
    fun getMessage(user: User, messagePath: String): Optional<String> {
        val language = user.playerLanguage

        return getMessage(user, messagePath, language)
    }

    /**
     * Gets a message from the messages file for the specified user's language.
     *
     * @param user        The user to get the message for
     * @param messagePath The path to the message in the messages file
     * @return An Optional containing the message, or empty if not found
     */
    fun getMessage(user: User, messagePath: String, language: String): Optional<String> {
        var messagePath = messagePath
        messagePath = "Messages.${messagePath}"


        var reader: ConfigReader
        try {
            reader = getConfigReader(language)
        } catch (exception: FileNotFoundException) {
            ServerSystem.log.log(Level.WARNING, "Failed to load messages for language '${language}': ${exception.message}", exception)
            reader = getConfigReader(defaultLanguage!!)
        }

        val message = reader.getString(messagePath, null)

        if (message == null) {
            ServerSystem.log.warning("Message '${messagePath}' not found for language ${language}!")
            return Optional.empty<String?>()
        }

        return Optional.of<String?>(message)
    }

    /**
     * Gets the ConfigReader for the specified language.
     *
     * @param language The language to get the ConfigReader for
     * @return The ConfigReader for the specified language
     * @throws FileNotFoundException If the messages file couldn't be found
     */
    private fun getConfigReader(language: String): ConfigReader {
        val readerOptional = languageLoader.getMessageReader(language)
        return readerOptional.orElseGet(Supplier { languageLoader.getMessageReader(defaultLanguage!!).get() })
    }
}